src/libasr/codegen/llvm_utils.cpp

```diff
@@ -3408,17 +3429,73 @@ namespace LCompilers {
         // TODO: operators other than "<"
-        llvm::Value* inequality_holds = llvm::ConstantInt::get(context, llvm::APInt(1, 0));
-        for( size_t i = 0; i < tuple_type->n_type; i++ ) {
-            llvm::Value* t1i = llvm_utils->tuple_api->read_item(t1, i, LLVM::is_llvm_struct(
-                                              tuple_type->m_type[i]));
-            llvm::Value* t2i = llvm_utils->tuple_api->read_item(t2, i, LLVM::is_llvm_struct(
-                                              tuple_type->m_type[i]));
-            llvm::Value* res = llvm_utils->is_less_by_value(t1i, t2i, module,
-                                        tuple_type->m_type[i]);
-            inequality_holds = builder->CreateOr(inequality_holds, res);
+
+        /**
+         * Equivalent in C++
+         * For "<"
+         * 
+         * equality_holds = 1;
+         * inequality_holds = 0;
+         * i = 0;
+         * 
+         * while( i < a_len && equality_holds ) {
+         *     equality_holds &= (a[i] == b[i]);
+         *     inequality_holds |= (a[i] < b[i]);
+         * }
+         * 
+         */
+
+        llvm::AllocaInst *equality_holds = builder->CreateAlloca(
+                                                llvm::Type::getInt1Ty(context), nullptr);
+        LLVM::CreateStore(*builder, llvm::ConstantInt::get(context, llvm::APInt(1, 1)),
+                          equality_holds);
+        llvm::AllocaInst *inequality_holds = builder->CreateAlloca(
+                                                llvm::Type::getInt1Ty(context), nullptr);
+        LLVM::CreateStore(*builder, llvm::ConstantInt::get(context, llvm::APInt(1, 0)),
+                          inequality_holds);
+
+        llvm::AllocaInst *idx = builder->CreateAlloca(llvm::Type::getInt32Ty(context), nullptr);
+        LLVM::CreateStore(*builder, llvm::ConstantInt::get(
+                                    context, llvm::APInt(32, 0)), idx);
+        llvm::BasicBlock *loophead = llvm::BasicBlock::Create(context, "loop.head");
+        llvm::BasicBlock *loopbody = llvm::BasicBlock::Create(context, "loop.body");
+        llvm::BasicBlock *loopend = llvm::BasicBlock::Create(context, "loop.end");
+
+        // head
+        llvm_utils->start_new_block(loophead);
+        {
+            llvm::Value* i = LLVM::CreateLoad(*builder, idx);
+            llvm::Value* cnd = builder->CreateICmpSLT(i, llvm::ConstantInt::get(
+                                context, llvm::APInt(32, tuple_type->n_type)));
+            cnd = builder->CreateAnd(cnd, LLVM::CreateLoad(*builder, equality_holds));
+            builder->CreateCondBr(cnd, loopbody, loopend);
+        }
+
+        // body
+        llvm_utils->start_new_block(loopbody);
+        {
+            llvm::Value* i = LLVM::CreateLoad(*builder, idx);
+            // llvm::Value* t1i = llvm_utils->tuple_api->read_item(t1, i, LLVM::is_llvm_struct(
+            //                                 tuple_type->m_type[i]));
```

Comment:
How do we index `tuple_type->m_type` with run time value `i`?

Reply:
We can't. :-). Tuples should only be indexed with fixed indices available at compile time so that type of the indexed item can be figured out at compile time itself. Its the pattern as C++ `std::tuple`.

Reply:
I see. Here we need to break out of the loop when equality does not hold at the first time. Run time values seem necessary for this.